### DISCUSSION TITLE
fhirlite architecture

### DISCUSSION THREAD

**Discussion by Dokotela** [fhir-fli/fhirlite#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. And, specifically for this discussion, I'm going to be trying to write my thoughts down about the internal infrastructure of what I'd like to accomplish (will somewhat mirror the [kanban in Projects](https://github.com/orgs/fhir-fli/projects/1). Please feel free to:
  * Join our [Slack Channel](https://join.slack.com/t/fhir-fli/shared_invite/zt-1sjsu9n93-JM8S4NLtSGCO6PQJmhu7ag)!
  * Email me: grey@fhirfli.dev
  * Share any of your own thoughts, suggestions, or ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by Dokotela** [comment]
So, first things are to probably re-summarize what my goals for this project are (and please feel free to add your own, ideally this can become a community project). This is an image I've been using for years. And I think it demonstrates where my aim is. There are lots of good data collecting apps, and some workable (although I wouldn't call any of them good), Electronic Health Records. I will give a shout-out to [OpenSRP](https://smartregister.org/) for building the kind of thing I'm interested in. But for a couple of reasons I could mention later, still not exactly what I envision. 

![image](https://user-images.githubusercontent.com/5497678/230410734-08c3653b-3457-4a16-8f82-8119943da475.png)

**Comment by Dokotela** [comment]
So that's my perception at least. I think there's a role that Flutter could play extremely well. That is a mobile-first, clinician- and patient-first, modular, scalable, and interoperable, EHR Lite platform. And what kind of architecture would I use to structure it? That's an excellent question, thank you for asking. 

There are two directions this can go, and I think they're important to discuss because while much of the underlying app architecture is the same, the backend architecture is slightly different. They are clinician vs patient facing. I'll be honest, while adminsitration is important, I don't care about it nearly as much (and I honestly don't give a f*ck about billing, so I will never be focusing on any modules for it, although if others would like to, please feel free).

## Clinician
I'm biased, so this is where I'm going to start, because I think it's usually done so terribly.

**Comment by Dokotela** [comment]
## FHIR Database
You obviously need a database. As a clinician, you'll need a central database for your organization. There are lots to choose from. There are a number of boxes that my ideal database would check:
1. Ideally this would be something that would run in the cloud, but you COULD run locally if you had to (especially for some settings where there is no internet or you have to keep your data in-country). 
2. Smoothly integrate with Flutter
3. Have real-time database capabilities
4. Full FHIR database functionality
5. Someone else who has done all the work so I don't have to set it up myself

### Classic FHIR Database
This one is hard, because I think there are a lot of options out there that _COULD_ be used, but aren't exactly what I want. And since I'm narcissistic, I'm going to talk about what I want.
* [HAPI](https://hapifhir.io/) - one of the first and probably best known FHIR backend servers. Ready and in use for production with their [commercial arm](https://www.smilecdr.com/). They also have one of the biggest [public FHIR servers](http://hapi.fhir.org/). Has essentially all functionality you'd want in a FHIR server, *EXCEPT* simple real-time capabilities (this is a big thing that I want, and probably my biggest complaint with most backends available). 
    * [MELD](https://meld.interop.community/) Just want to mention MELD because I like them and have worked with them. It's a free HAPI sandbox that you can signup and use. Maintained by some folks in Michigan, they're very friendly and helpful, so if you'd like to play around with it, I can recommend trying here.
* [Aidbox](https://www.health-samurai.io/aidbox) I don't think Aidbox always gets the recognition they deserve. They have a really smooth server, all built on Postgres (plus they all do it in Clojure, who even remembers how to write anything in Lisp?). Anyway, I wish I could look into how they do some of their Postgres work (they publish a lot as open-source, but obviously they're trying to succeed so some of it remains closed), because I like Postgres as a database.
* [Google's Healthcare API](https://cloud.google.com/healthcare-api) So all of the big players have something equivalent. I just happen to like Google's Cloud and have worked with it, so I'm going to post it here (but Azure, AWS, IBM, Cloudforce all have something vaguely equivalent). Probably the most easily scalable of the above 3, plus super easy to spin-up a new instance. As all of them, however, *NOT REAL-TIME*.

### Not FHIR Specific Database
So these are some of the backends available that work smoothly with Flutter, and would be great if I could somehow merge them into FHIR.
* [Firestore](https://firebase.flutter.dev/docs/firestore/usage/) not necessarily the best, but probably the original. **NOTE:** this is Firestore **NOT** Firebase. Firebase is not HIPAA compliant. Anyone who has done Flutter dev is probably aware of Google's managed, NoSQL/Document database. It does a lot of things really nicely, including: real-time updates, lots of monitoring, authentication. However, it doesn't have any normal FHIR integration, you have to setup anything manual if you want to keep past versions of resources, and there's no built-in FHIR searches the way the FHIR databases above do.
    * One thought I had that I did want to bring up. You _CAN_ use triggers on updates in Firestore. I think you could use these triggers to write the updated Resource to Google's Healthcare API. That would then mean you could have an updated, and accurate, live FHIR database in firestore, that was mirrored (including versions) in the Healthcare API, that you could also search and use for other FHIR functionality. The downside of course would be that syncing wouldn't be both ways (you couldn't updated Healthcare API and have it update Firestore...I don't think)
* [Supabase](https://supabase.com/) A new one I've been playing around with and I have to say it works really nicely. Again, it has gone out of it's way to include a lot of Flutter and Dart integration, so that part is easy, again, also supports authentication, monitoring, and real-time updates. Plus it's open-source _AND_ you could host it yourself. It's built on Postgres, so all of the Postgres goodness that you know and love you could use with this too. If Aidbox ran all of their Postgres magic on Supabase, this would almost certainly be my go-to backend. But alas, they haven't yet, which means I would have to figure out all of the Postgres functionality myself. Not my forte, so if someone else is interested, that would be great.
    * One **EXTREMELY** important note, Supabase is currently not HIPAA compliant (although they are [working on it](https://supabase.com/blog/supabase-enterprise)). So if you're working outside of the US it's probably ok, but in the US you'd have to do some work on your own to ensure compliance.  
* [Appwrite](https://appwrite.io/) another really handy backend that's open-source. I haven't played with this much, but they again appear to have done a lot of work to integrate it with Flutter (amongst other offerings). It does do real-time, authentication, and analytics. It's probably the most open of these three, which means it's the most flexible, and the one you have to do the most to support. 
    * I'm including this because I have a friend who likes it, but it doesn't technically meet my full specifications, because it doesn't come with a native Database backend. So that requires you to run the database and sit this on top of it. 
* [Serverpod](https://serverpod.dev/) - another I haven't used, but always want to give a shoutout to anyone who uses Dart. This is similar to the above, a server that handles a lot of things for you but **NOT** the database. Still, it sounds cool.

## Bottom Line
From the pure FHIR servers, I like Google's Healthcare API. But for this particular project, I think my recommendation is going to be Supabase (and possibly Firestore). But as mentioned, if I could somehow sit Supabase on Aidbox, then that would be it and done.

**Comment by Dokotela** [comment]
## Proxy Server (technically Reverse Proxy)
This was not something that I had thought about until recently. But realistically in order to be secure and try and protect healthcare data appropriately, you should probably have a Proxy server sit in front of your database. You're welcome to use any that you want, I don't know that one is truly better than another. I will give to mentions here given my above recommendations:
1. Apigee X - if you're going to do anything with Google Cloud, Apigee is probably the simplest solution for a proxy server.
2. NGINX - the most popular web server, can also be used as a reverse proxy. In fact, there's this handly how-to in order to [use it with Supabase](https://www.linode.com/docs/guides/installing-supabase/). 

This is an area where I'm not particularly knowledgeable, so feel free to add your thoughts.

**Comment by Dokotela** [comment]
## Data Model
If you're here there should be no argument about this. But just in case...it's FHIR.

FHIR is one of the few data models in healthcare that has been federally endorsed **AND REQUIRED** by the US Government. This means you're unlikely to find any other clinical data model that is going to receive the same level of money and support, that alone is probably enough to make it the recommendation.

Interoperability. So you're in the US, and you want to interface with Epic, Cerner, Allscripts, you need FHIR. But, this project was not initially envisioned for us in the US. Which means those don't apply. And I've certainly heard the suggestion that the OpenEHR data model is better for storage. Even if this is true (and I'm not convinced it is), OpenEHR means you have to interface with OpenEHR. But what if you want to interface instead with OpenMRS? Or you DO want to move your data to the US? Ultimately we're talking a mobile platform, so you may not even have control of the backend. This almost insists that you use FHIR.

For international work, you could argue, maybe, that DHIS2 is better, or more widely used. In Africa that may be the case, but I don't think you can argue that for the rest of the world. Not to mention, I've spoken to one of the heads of DHIS2 in Africa, and he told me that he has **NEVER** seen it used to entire data at point of care. That's not to say that DHIS2 is a bad data model, but not one that I think fits our use case. Also, just to make note, here is some [DHIS2 work to convert to FHIR](https://dhis2.org/integration/fhir/), and here is [Ona's take on it](https://ona.io/home/powering-interoperability-with-a-fhir-to-dhis2-adapter/).

I don't think anyone is going to argue to use OMOP for clinical care, but this paper goes even further and says why bother transforming data to OMOP, even if it's easier to analyze for research, it's too much extra work, [just use FHIR instead](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8050750/).

So yeah...FHIR.

**Comment by Dokotela** [comment]
## Mobile Framework
Again, if you're here, this isn't an argument. 

If you ever need to argue with someone, feel free to Google why use Flutter, there's lots of better worded arguments than I'll make here. The only real competitor is React Native, and this graph gives at least one good answer why it's not the right choice.

![image](https://user-images.githubusercontent.com/5497678/230472667-70e5e39a-d7fc-4d0c-b7c5-8faf12eaf584.png)

But like I said, if you're here, I don't have to convince you, it's Flutter.

**Comment by Dokotela** [comment]
## Authentication
The starting point for the user of an application. I think we have to support all of the usual culprits:
- Username/password
- Biometrics (not necessary, but so convenient)
- Social login (probably only Google and Microsoft, unless others specifically requested)
- Smart on FHIR (obviously)
- [@sign](https://atsign.com/) This one is interesting, and I'm not sure the best way to use it. Take a look at the work they've done, it's phenomenal, and they're really great guys.  I'm not sure how useful this would be for an organization, but I may come back to it for the patient facing apps.

**Comment by Dokotela** [comment]
## Local Storage
So we all know (I hope) that storing things on a mobile device is generally discouraged for security reasons, especially with healthcare data. However, we also know that especially for this project, there are times we need to save things locally. These are my thoughts about it:

- Required Locally encrypted (At least AES-128 - could also use something like Salsa20)
- Already have SQFLite/Sembast
- Could consider SQLCipher, Hive, Isar or Objectbox for the future
- Locally encrypted option for Firestore if going that route (again, another reason why Firestore is useful, despite missing features)

Future Questions:

- There may be times when a single device is used by multiple users, especially if out of range of the internet for multiple days in a row. 
- If this is the case, can they all access that device? 
- Or should each user have a separate database that only they can access on that device?
- If you're going to allow multiple people to access the data on that device, how and where are you going to define what they have access to?
- How/where do you store an encryption key if more than one person can have access to it?

**Comment by Dokotela** [comment]
## Data sync, reporting, and transport
FHIR is all JSON, so should easily be exportable to anything that accepts it.

- RESTful interactions with most of the above databases is already available
- Realtime interactions with Firestore or Supabase should be created (I'm hoping to get around to it sooner, rather than later)

Future Considerations
- For low resource settings, should there be an option to export as a flat file?
- If so, should we transform it to a CSV? Most FHIR resources, while in theory could have deep nesting, usually doesn't, and we could probably take the most used fields and convert it to a spreadsheet.
- This is more a FHIR than a FHIR Lite question, but providing an easy way to get data in and out of the system in easy formats seems pertinent for this project.

**Comment by Dokotela** [comment]
## Communications
Again, 2 versions:
1. In-app between providers
2. Possibly in-app, but more likely out of app between Providers and patrons

There are a number of ways to do this, although I'm not familiar with most of them. Again, I would lean towards the [@protocol](https://atsign.com/) for this one, since I think there are other uses of it, and I'd like to take advantage of them when possible. 

However, there will need to be ways to communicate with patrons that probably do not rely on in-app communication (as many patrons won't have the app). That probably requires either sms or whatsapp. Some of this could be done directly, or through Twilio, but I don't think that integrates well into an app. 

### TODO: this is an area that could use more thought.

**Comment by Dokotela** [comment]
## Modules
Alright, I think that's most of the basic architecture decisions that need to be made. Obviously, any and all thoughts, inputs or suggestions on any of this is more than welcome. 

But this post is about modules. The benefit of modern architecture, and Flutter specifically is the ease with which we should be able to add and remove functionality by dropping in another package anytime we like with that functionality. So, what kind of functionality am I thinking? This will be an ongoing list, but here is some functionality that the app should at least have the option of providing:

### Required modules (at least eventually)
- Scheduling
- Growth
- Immunizations
- Consent
- Automated date/time/GPS recordings
- Ability to print
- Ability to generate pdf files
- Creates graphs (should have a quick and easy way to use one of the graph packages in Flutter to create graphs)
- Multi-lingual (this one is done, we have an easy way of building multiple translations into either an application or FHIR resources)
- Ability to capture pictures

### Desired modules (not 100% necessary, but convenient)
- Dictation
